AnyVisionltd/automation-infra
# Hab/heartbeat

** What does this PR do? **

Adds initial heartbeat implementation from test infrastucture 
Creates a dedicated heartbeat service
Updates Allocate so that it expires expired jobs
Sends resourcemanager updates about expired jobs
Adds a stubbed out 'teardown' method in resource manager
Adds logic to still_free so that it actually checks to see if a job is still free (support multiple resources)
Tidies up a few things (linting et al). Tidy up as I go ..

** How should I test this? **

Spin all the services up and run a test. You should see:

Heartbeat output:
```
[L:31]# DEBUG    [31-03-2020 19:08:38]  expiration extended
[L:17]# DEBUG    [31-03-2020 19:08:41]  received heartbeat from allocation_id f2cc3538-0ab7-4d21-9e48-de35277259d5
[L:31]# DEBUG    [31-03-2020 19:08:41]  expiration extended
[L:17]# DEBUG    [31-03-2020 19:36:45]  received heartbeat from allocation_id 8b7232ad-924c-4ba0-9d61-d6ffb4d52ac7
[L:31]# DEBUG    [31-03-2020 19:36:45]  expiration extended
```
Allocate output:
```
[L:15]# DEBUG    [31-03-2020 19:37:15]  processing expired jobs
[L:15]# DEBUG    [31-03-2020 19:37:16]  processing expired jobs
[L:21]# DEBUG    [31-03-2020 19:37:16]  found expired job! processing
[L:22]# DEBUG    [31-03-2020 19:37:16]  {'state': 'claimed', 'allocation_id': '8b7232ad-924c-4ba0-9d61-d6ffb4d52ac7', 'demands': {'benchmark_machine': {'cpu': '>=1'}}, 'resourcemanager_id': 'b0c364a0-83b4-424a-844b-6be58c5c8a84', 'inventory_type': 'static', 'inventory_ref': 'virtualboxubuntu', 'expiration': 1585683435.7512004}
[L:33]# DEBUG    [31-03-2020 19:37:16]  notified dedicated resource queue
[L:35]# DEBUG    [31-03-2020 19:37:16]  deleted job
```
Resource Manager output:
```
[L:43]# DEBUG    [31-03-2020 19:36:45]  succeeded
[L:30]# DEBUG    [31-03-2020 19:37:16]  {"expired_job": {"state": "claimed", "allocation_id": "8b7232ad-924c-4ba0-9d61-d6ffb4d52ac7", "demands": {"benchmark_machine": {"cpu": ">=1"}}, "resourcemanager_id": "b0c364a0-83b4-424a-844b-6be58c5c8a84", "inventory_type": "static", "inventory_ref": "virtualboxubuntu", "expiration": 1585683435.7512004}}
[L:53]# DEBUG    [31-03-2020 19:37:16]  tearing down resource
```
